editor.planx.uk/src/pages/FlowEditor/lib/analyticsProvider.tsx

```diff
@@ -491,17 +491,50 @@ export const AnalyticsProvider: React.FC<{ children: React.ReactNode }> = ({
   }
 
   function getAllowListAnswers(
+    nodeId: string,
     breadcrumb: Store.userData,
-  ): Record<string, any>[] | undefined {
+  ): Record<string, unknown>[] | undefined {
+    const answers = getAnswers(nodeId);
+    const data = getData(breadcrumb);
+
+    const allowListAnswers = [...answers, ...data];
+    if (!allowListAnswers.length) return;
+
+    return allowListAnswers;
+  }
+
+  /**
+   * Extract allowlist answers from user answers
+   * e.g. from Checklist or Question components
+   */
+  function getAnswers(nodeId: string) {
+    const { data } = flow[nodeId];
+    const nodeFn: string = data?.fn || data?.val;
+    if (!nodeFn || !ALLOW_LIST.includes(nodeFn)) return [];
+
+    const answerIds = breadcrumbs[nodeId]?.answers;
+    if (!answerIds) return [];
+
+    const answerValues = answerIds.map((answerId) => flow[answerId]?.data?.val);
+
+    // Match data structure of `allow_list_answers` column
+    const answers = [ {[nodeFn]: answerValues } ];
```

Comment:
Having a brief look at the `analytics_logs` I think as the data  structure there'll now be a mismatch between current and old data i.e. array of values vs array of objects:

<img width="655" alt="Screenshot 2024-03-01 at 19 03 45" src="https://github.com/theopensystemslab/planx-new/assets/36415632/ac3d28ae-8b7d-48e4-a3c3-74c39deff126">
<img width="843" alt="Screenshot 2024-03-01 at 19 04 20" src="https://github.com/theopensystemslab/planx-new/assets/36415632/fd4ca57f-ab3a-43e6-9563-7b74af3699f6">


Although seeing as we haven't been tracking `allow_list_list` I wonder if we can just crack on? 

Otherwise I think it would be possible to update historical data using the `node_fn` value as the key 🤔

Reply:
Yep - good point. First thought would be to ignore this as it's a new column, however having data in different formats in the column would prove tricky to query so we might as well do this properly!

There's more logs populated than expected already on production, so I think clearing older logs is not really a solution here.

I've tested the following SQL locally which will convert the column. Could you please review and manually test the SQL below @Mike-Heneghan before we run on production (once this change has gone to prod).

To select rows with the "old" data format - 

```sql
SELECT
    id,
    node_fn,
    node_type,
    allow_list_answers
FROM
    analytics_logs
WHERE
    jsonb_typeof(allow_list_answers) = 'array'
    AND jsonb_typeof(allow_list_answers -> 0) = 'string';
```

And then this to update "old" → "new" - 

```sql
UPDATE analytics_logs
SET allow_list_answers = (
    SELECT jsonb_agg(jsonb_build_object(key, value))
    FROM (
        SELECT 
            node_fn AS key, 
            jsonb_array_elements_text(allow_list_answers) AS value
        FROM analytics_logs
        WHERE jsonb_typeof(allow_list_answers) = 'array' AND 
              jsonb_typeof(allow_list_answers->0) = 'string'
    ) AS subquery
)
WHERE jsonb_typeof(allow_list_answers) = 'array' AND 
      jsonb_typeof(allow_list_answers->0) = 'string';
```

Reply:
Hey Daf, I think there's something not quite right based on my local testing. 

The `SELECT` works as expected: 

<img width="1936" alt="Screenshot 2024-03-04 at 12 07 38" src="https://github.com/theopensystemslab/planx-new/assets/36415632/c6edb79c-a471-4667-90f7-b4d6557534db">

Although fetching the same records again post `UPDATE`:

<img width="1936" alt="Screenshot 2024-03-04 at 12 10 54" src="https://github.com/theopensystemslab/planx-new/assets/36415632/1bcdfa45-628f-4057-a38a-4df718a93577">

It appears it doesn't group the values against the key and also looks like every record has become an amalgamation of the rest?

Reply:
Ha! That doesn't match my results - I maybe just had a simple use case 😅 Let me take another look!

Reply:
No worries! This sort of stuff is really fiddly, please let me know when you want me to retest 👍

Reply:
@Mike-Heneghan Could you please try out this query?

```sql
UPDATE analytics_logs
SET allow_list_answers = (
    SELECT jsonb_agg(jsonb_build_object(node_fn, subquery.values))
    FROM (
        SELECT jsonb_agg(value) AS values
        FROM (
            SELECT 
              node_fn,
              jsonb_array_elements_text(allow_list_answers) AS value
            FROM analytics_logs
            WHERE 
              jsonb_typeof(allow_list_answers) = 'array' AND 
              jsonb_typeof(allow_list_answers->0) = 'string'
        ) AS subquery_inner
    ) AS subquery
)
WHERE 
  jsonb_typeof(allow_list_answers) = 'array' AND 
  jsonb_typeof(allow_list_answers->0) = 'string';
```

I'm getting the expected results here - 

![image](https://github.com/theopensystemslab/planx-new/assets/20502206/8ca34270-8ba8-4d61-96bd-e488ed35547f)

Reply:
It seems like this maybe has the same issue based on my testing 😅 

<img width="1577" alt="Screenshot 2024-03-06 at 11 46 18" src="https://github.com/theopensystemslab/planx-new/assets/36415632/fbdff746-dbaf-44dd-acc9-775c1eb39d5c">

<img width="1577" alt="Screenshot 2024-03-06 at 11 47 40" src="https://github.com/theopensystemslab/planx-new/assets/36415632/2925df43-f02b-4aff-938a-58172a0fa48e">

Reply:
Ah I get it sorry - I'm just testing on a single record 🤦 

Probably just needs and id=id clause. Taking a look and thanks for testing!

Reply:
Hey sorry for the slow turnaround here @Mike-Heneghan 

Clause added and highlighted with comment - 

```sql
UPDATE analytics_logs
SET allow_list_answers = (
    SELECT jsonb_agg(jsonb_build_object(node_fn, subquery.values))
    FROM (
        SELECT jsonb_agg(value) AS values
        FROM (
            SELECT 
              node_fn,
              jsonb_array_elements_text(allow_list_answers) AS value
            FROM analytics_logs inner_log
            WHERE 
              -- New clause to match ids
              inner_log.id = analytics_logs.id AND
              jsonb_typeof(inner_log.allow_list_answers) = 'array' AND 
              jsonb_typeof(inner_log.allow_list_answers->0) = 'string'
        ) AS subquery_inner
    ) AS subquery
)
WHERE 
  jsonb_typeof(allow_list_answers) = 'array' AND 
  jsonb_typeof(allow_list_answers->0) = 'string';
```

Getting expected results here, but as always would love another pair of eyes 😄

Reply:
Working as expected for me ✅ 

I regret not adding an order by to my query used to check results 😅 

Although, comparing the updates are made as expected for me and it doesn't affect the data already in the new structure 🥳 

Before:

<img width="1957" alt="Screenshot 2024-03-13 at 12 14 58" src="https://github.com/theopensystemslab/planx-new/assets/36415632/438495bc-62ef-4079-a38d-35017ab9a7de">

After:

<img width="1957" alt="Screenshot 2024-03-13 at 12 15 44" src="https://github.com/theopensystemslab/planx-new/assets/36415632/bd4b63ae-dbb4-499b-8a46-005b95da25ad">

Reply:
Ah perfect - honestly thanks again for being so on-point with testing here, really helped! 🙌